src/PointEvalHandler.jl

```diff
@@ -89,54 +91,70 @@ function _get_cellcoords(points::AbstractVector{Vec{dim,T}}, grid::AbstractGrid,
-function find_local_coordinate(interpolation, cell_coordinates::Vector{<:Vec{dim}}, global_coordinate::Vec{dim}; tol_norm = 1e-10) where dim
+function find_local_coordinate(interpolation::Interpolation{ref_shape}, cell_coordinates::Vector{<:Vec{sdim}}, global_coordinate::Vec{sdim};  warn::Bool = false, linesearch_max_substeps::Int = 10, max_iters::Int = 10, tol_norm = 1e-10) where {sdim, ref_shape}
     T = promote_type(eltype(cell_coordinates[1]), eltype(global_coordinate))
     n_basefuncs = getnbasefunctions(interpolation)
     @assert length(cell_coordinates) == n_basefuncs
-    local_guess = zero(Vec{dim, T})
-    max_iters = 10
+    local_guess = cellcenter(ref_shape, T)
     converged = false
-    for _ in 1:max_iters
-        global_guess = zero(Vec{dim, T})
-        J = zero(Tensor{2, dim, T})
-        # TODO batched eval after 764 is merged.
-        for j in 1:n_basefuncs
-            dNdξ, N = reference_shape_gradient_and_value(interpolation, local_guess, j)
-            global_guess += N * cell_coordinates[j]
-            J += cell_coordinates[j] ⊗ dNdξ
-        end
+    for iter in 1:max_iters
+        # Setup J(ξ) and x(ξ)
+        mapping, global_guess = calculate_mapping_and_spatial_coordinate(interpolation, local_guess, cell_coordinates)
+        J = getjacobian(mapping)
+        # Check if converged
         residual = global_guess - global_coordinate
-        if norm(residual) <= tol_norm
-            converged = true
+        best_residual_norm = norm(residual) # for line search below
+        # Early convergence check
+        if best_residual_norm ≤ tol_norm
+            converged = check_isoparametric_boundaries(ref_shape, local_guess, sqrt(tol_norm))
+            break
+        end
+        if calculate_detJ(J) ≤ 0.0
+            warn && @warn "det(J) negative! Aborting! $(calculate_detJ(J))"
+            break
+        end
+        Δξ = _solve_helper(J, residual) # J \ b throws an error. TODO clean up when https://github.com/Ferrite-FEM/Tensors.jl/pull/188 is merged.
+        # Do line search if outside element
+        best_index = 1
+        new_local_guess = local_guess - Δξ
+        global_guess = spatial_coordinate(interpolation, new_local_guess, cell_coordinates)
+        best_residual_norm = norm(global_guess - global_coordinate)
+        if !check_isoparametric_boundaries(ref_shape, new_local_guess, sqrt(tol_norm))
+            for next_index ∈ 2:linesearch_max_substeps
+                new_local_guess = local_guess - Δξ/2^(next_index-1)
+                global_guess = spatial_coordinate(interpolation, new_local_guess, cell_coordinates)
+                residual_norm = norm(global_guess - global_coordinate)
+                if residual_norm < best_residual_norm && check_isoparametric_boundaries(ref_shape, new_local_guess, sqrt(tol_norm))
+                    best_residual_norm = residual_norm
+                    best_index = next_index
+                end
+            end
+        end
```

Comment:
So long since we discussed this that I don't recall, but do you have an example which triggers this block? Would be nice to comment on when this could happen.

Reply:
The obvious one is (as indicated in the comment above) if you are close to the element boundary the line searcher gets triggered, so we can also support point finding for problematic cases where the evaluation of the ansatz functions outside the element gives you something unexpected. This also gets triggered if the Newton overshoots and you would move outside the element.

Reply:
Let me change it to a projected Newton then, makes probably more sense.

Reply:
Aha, thanks for explaining! 
So it is only to prevent getting outside? 
In that case, wouldn't it suffice to only check this during line-searching to simplify the code? E.g. 
```suggestion
        # If we are outside the element, reduce the step length until we are inside again
        while !check_isoparametric_boundaries(ref_shape, new_local_guess, sqrt(tol_norm)) && best_index < linesearch_max_substeps
            best_index += 1
            new_local_guess = local_guess - Δξ/2^(best_index-1)
        end
        local_guess = new_local_guess
        global_guess = spatial_coordinate(interpolation, new_local_guess, cell_coordinates)
        best_residual_norm = norm(global_guess - global_coordinate)
```
? 
To me, it looks like the current code will always run `linesearch_max_substeps`, and then just take the step with the lowest residual?

Reply:
> To me, it looks like the current code will always run linesearch_max_substeps, and then just take the step with the lowest residual?
The one with the lowest residual, which is still inside the element.

Since I am not sure about the interaction with the convergence for your algorithm I think we should just stick to the normal Newton line searcher, which has linear convergence on the element boundary and essentially quadratic convergence inside the element. I think if we invest a bit more time we can even write down a proof that it the line searcher converges.

Reply:
I'm not suggestion to change the update algorithm, just the stopping criterion. 
But if you don't agree, then let's keep it and I can do a follow up PR to simplify at some point.